### PR TITLE
groupcheck, fwupd, fwupdate

### PR DESCRIPTION
**Pull request by pohly**
This PR makes groupcheck usable in Refkit. fwupd depends on it for
authorizing the "fwupdmgr update" operation.

fwupd was tested manually against a Logitech Unifying Receiver. See
the fwupd/README for details. There is no automated QA check, nor does
fwupd get built by default at the moment (not added to any image).

fwupdate gets packaged and fwupd is configured to use it, but could
not be tested at all due to the lack of a UEFI firmware with UEFI
capsule support.

Please provide feedback on how to proceed with QA. We could either add
fwupd to some image (which one?) or add it to the CI build
targets. Enabling fwupd-demo-firmware would be nice for testing, but
then we end up adding the non-redistributable firmware to our
download cache and images. Those are not public, but even internally I
am not sure whether we would have the rights to do that.

As far as I am concerned, that's all that I can do for
https://bugzilla.yoctoproject.org/show_bug.cgi?id=11468 at the moment.

### CODE REVIEW

**Comment by ipuustin** [comment]
You might want to use groupcheck 2.0 release tag or tarball: https://github.com/ostroproject/groupcheck/releases. Maybe refkit should ship its own groupcheck.service file instead of patching the example one? Other than that the groupcheck changes look good to me.

**Comment by pohly** [comment]
On Mon, 2017-09-18 at 06:54 +0000, Ismo Puustinen wrote:
> You might want to use groupcheck 2.0 release tag or tarball:
> https://github.com/ostroproject/groupcheck/releases.

tarball would work, but I prefer git for now simply because there might
be further experimental changes and then git is a bit more usable.

Using the tag won't work well because it implies remote access to the
repo to get the current tag for each build. The actual code right now
is exactly the 2.0 version.

I'll add PV="2.0" to make that clear.

> Maybe refkit should ship its own groupcheck.service file instead of
> patching the example one?

That would imply copying quite a few things, including the user setup.
Not sure which solution is better overall.

Perhaps upstream groupcheck could take a "--with-systemd-config=-d
/usr/share/groupcheck.d" configure parameter and then embed that into
the final systemd unit file?

**Comment by pohly** [comment]
test this please

Unrelated test failure (570x.refkit-image-computervision.oeqa.runtime.programming.python.apprt_python.SanityTestPython.test_python_helloworld).

**Comment by pohly** [comment]
@mythi, @ipuustin: I had to add some checks that prerequisites of fwupd are actually present. I ended up creating "check-available.bbclass" and used that to also simplify some other recipes. Please review.

Whether to build and test fwupd is still an open question, though.

**Comment by pohly** [comment]
@ipuustin nftables tests failed in https://intel-iot-refkit.01.org/job/intel-iot-refkit_pull-requests/569/ - is that something that could have been broken by this PR? I can't imagine why.

**Comment by ipuustin** [comment]
I've seen the same nftables error before. I was thinking there might be a timing issue in the tests (or rather in the framework which executes the tests), since the problem appears to be non-deterministic. Nftables itself should be pretty solid. I don't think it's related to fwupd. test this please

**Comment by samniem** [comment]
test this please

**Comment by samniem** [comment]
test this please

File: meta-refkit-core/bbappends/meta-intel/common/recipes-kernel/linux/linux-intel_%.bbappend [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FILESEXTRAPATHS_prepend := "${THISDIR}/files:"
+
+# Long-term solution has to be to get this into the default OE
+# kernel configs, see "[OE-core] hidraw support in the kernel(s)".
+SRC_URI_append_df-refkit-config = " file://hidraw-class.cfg"
```
[the remark this answers is not in the document]
Comment (pohly) [reply meta-refkit-core/bbappends/meta-intel/common/recipes-kernel/linux/linux-intel_%.bbappend]: No, not yet. Bruce's reply left too many questions open for me. 

Having looked at that file now, it seems like the wrong place. It seems to be about sensors, which might be disabled on systems which only support USB mice and keyboards.

File: meta-refkit-core/recipes-extended/fwupdate/fwupdate_9.bb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+LIC_FILES_CHKSUM = "file://${S}/COPYING;md5=6d7bfc784f89da02599036c034adfb34"
+DEPENDS = " \
+    efibootmgr \
+    efivar \
+    gnu-efi \
+"
+
+SRC_URI = " \
+    git://github.com/rhboot/fwupdate.git;method=https \
+    file://libfwup.c-fix-uninitialized-memory-read-in-boot-orde.patch \
```
[the remark this answers is not in the document]
Comment (pohly) [reply meta-refkit-core/recipes-extended/fwupdate/fwupdate_9.bb]: Upstream fixed the problem independently. I can replace with the backported patch, if you want.

File: meta-refkit-core/bbappends/meta-intel/common/recipes-kernel/linux/linux-intel_%.bbappend (hunk 1 above)
[the remark this answers is not in the document]
Comment (mythi) [reply meta-refkit-core/bbappends/meta-intel/common/recipes-kernel/linux/linux-intel_%.bbappend]: True. Anyway my preference is we keep the goal of not having ```linux-intel_%.bbappend```.

File: meta-refkit-core/recipes-extended/fwupdate/fwupdate_9.bb (hunk 2 above)
[the remark this answers is not in the document]
Comment (mythi) [reply meta-refkit-core/recipes-extended/fwupdate/fwupdate_9.bb]: Or just add ```Upstream-Status: Inappropriate [reason]```

File: meta-refkit-core/bbappends/meta-intel/common/recipes-kernel/linux/linux-intel_%.bbappend (hunk 1 above)
[the remark this answers is not in the document]
Comment (pohly) [reply meta-refkit-core/bbappends/meta-intel/common/recipes-kernel/linux/linux-intel_%.bbappend]: I've removed the linux-intel_%.bbappend. As it is unclear whether it should be added to the default kernel, I now only document the need for it for USB devices in meta-refkit-core/recipes-extended/fwupd/README.

**Comment by pohly** [comment]
@mythi the revised commits should address the review feedback, please have another look.

**Comment by pohly** [comment]
@mythi there's just one "minor" issue left: we don't even build fwupd + fwupdate at the moment. Shall I add them to the CI target list to fix that?

**Comment by pohly** [comment]
test this please

I've added fwupd as build target, to ensure that it really builds. Building locally revealed that the fwupdate patches had to be revised some more.

**Comment by mythi** [comment]
On Fri, Sep 29, 2017 at 10:27 PM, Patrick Ohly  wrote:

> test this please
>
> I've added fwupd as build target, to ensure that it really builds.
> Building locally revealed that the fwupdate patches had to be revised some
> more.
>
>
>
Good that you added it! It has spotted even more problems; fwupdate is also
failing (expr syntax error?)

**Comment by pohly** [comment]
fwupdate was using the host's "objcopy", instead of the one from binutils-cross. That worked on my desktop, but failed on the CI system (objcopy too old there). Fixed with "OBJCOPY='${OBJCOPY}'" when invoking make.

**Pull request merged**